### PR TITLE
Enable `dim=None` for `torch.sum`

### PR DESCRIPTION
**Pull request by kurtamohler**
Part of #29137

### CODE REVIEW

**Comment by kurtamohler** [comment]
I'm having trouble figuring out this failure with `./build/bin/test_jit --gtest_filter=TestShapeGraphLinting.Basic`. Here's the error it gives:

```
C++ exception with description "mapped_type->isSubtypeOf(graph->inputs().at(i)->type()) INTERNAL ASSERT FAILED at
"/work2/kurtamohler/development/pytorch-dim-none/torch/csrc/jit/runtime/symbolic_shape_registry.cpp":176, please report a
bug to PyTorch. For schema type: int[]? Expected supertype of int[]? but got graph_type int[] at index 1 of schema:
aten::sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, int? dtype=None) -> (Tensor)
```

So evidently, when the JIT graph node for `sum` is created, its `dim` input is given the type `int[]`, but it needs to be `int[]?`. I'm not sure how to change that

**Comment by kurtamohler** [comment]
> I think the "shape function registry" entry needs to be updated:

I did change the `int[1]` on that line to `int[1]?`. But now I notice the `mean_dim` thing there, and that probably is the issue. I'll try making a separate `sum_dim` function

**Comment by kurtamohler** [comment]
Hey @JackCaoG, I'm getting an XLA failure because I changed the `at::IntArrayRef` arg for `sum` to `at::OptionalIntArrayRef`.

```
2022-05-05T02:53:53.1973363Z /var/lib/jenkins/workspace/xla/torch_xla/csrc/aten_xla_type.cpp:3263:32: error: out-of-line definition of 'sum' does not match any declaration in 'torch_xla::XLANativeFunctions'
2022-05-05T02:53:53.1974385Z at::Tensor XLANativeFunctions::sum(const at::Tensor& self, at::IntArrayRef dim,
```

**Comment by ngimel** [comment]
I'm not sure what's the best way to do this, maybe it's giving dims argument default value? See how it's done in `nansum`, that doesn't require introducing all Optional plumbing.

**Comment by JackCaoG** [comment]
Hey can you give me a ping when this pr is more ready? If it is just a `at::IntArrayRef` to optional change then it should be pretty quick. I am afraid there are more I need to change on the xla side as this pr going through reviews so I prefer to do it when this pr is more final.

**Review by ezyang — APPROVED** [review]
Solid stuff

**Comment by ezyang** [comment]
@JackCaoG this looks good to go, we're ready!

**Comment by JackCaoG** [comment]
Thanks, I will work on a fix on xla side.

**Comment by JackCaoG** [comment]
I will update here when all tests in https://github.com/pytorch/xla/pull/3554 passed

**Comment by ngimel** [comment]
Here's roughly what I'm talking about https://github.com/pytorch/pytorch/compare/ngimel/sum_overload?expand=1, it also adds `out=` support to the no-dim case that doesn't support it now. It of course requires a lot of refinement with looking at the places where current signature is hardcoded, but current PR does it too, and fewer overloads are better than more overloads. 
cc @ezyang

**Comment by JackCaoG** [comment]
Test passed on xla end, feel free give me a ping when this one landed.

**Comment by peterbell10** [comment]
@ngimel supporting `dim=None` as an all-reduce is required for full NumPy compatibility. `dim=[]` doing an all-reduce is not part of the NumPy behaviour so making it the default would be counter-productive.

**Comment by ngimel** [comment]
And yet () doing all-reduce is still a de-facto situation for most reductions, and changing () behavior for sum would be very disruptive.

**Comment by peterbell10** [comment]
Yes, deprecating `dim=()` would be disruptive. However, this PR doesn't change any existing behavior, it only adds the `dim=None` behavior which a step towards better NumPy compatibility.

**Comment by ezyang** [comment]
The lynchpin is that Numpy accepts an explicit None, so the defaulting trick is insufficient. @ngimel does this address your concerns?

**Comment by ngimel** [comment]
The defaulting trick is sufficient to accept `dim=None` (and `out`), it works on `nanmean` and `nansum`, whether we will indeed change the behavior of `()` is an open question, but in any case at that point it can be addressed for all reductions together, and all reductions having similar overloads will be helpful.

**Comment by ezyang** [comment]
Is the request to apply this change to the other reductions, in that case?

**Comment by peterbell10** [comment]
> The defaulting trick is sufficient to accept `dim=None` (and `out`), it works on `nanmean` and `nansum`

Okay, I didn't realise this behavior existed. If anything it looks more like a bug in `PythonArgParser` though. It's not reflected in the JIT schema, so TorchScript won't accept it:
```
@torch.jit.script
def nansum_script(x):
    return torch.nansum(x, dim=None)
```
fails to compile with the following error
```
Expected a value of type 'List[int]' for argument 'dim' but instead found type 'NoneType'.
```

**Comment by ezyang** [comment]
I agree with peterbell

**Comment by kurtamohler** [comment]
Natalia's suggestion is simpler to implement, but with the points Peter brought up, `int[1]? dim` still seems like the better option to me. What do you think @ngimel ?

**Comment by kurtamohler** [comment]
So should we go with `int[1]? dim` or `int[1] dim=[]` ?

**Comment by ezyang** [comment]
Natalia is on vacation right now, I can ping her again when she gets back

**Comment by ezyang** [comment]
Ok so we can ship this as is?

**Comment by ngimel** [comment]
Sure, we can, @JackCaoG what's the new process for xla fixes?

**Comment by JackCaoG** [comment]
@ngimel Let me rebase my xla pr and make sure everything still works. In terms of merge plan, in Brian's word

```
(1) update your core PR’s `xla.txt` file to point to a commit hash from your pt/xla PR
(2) After tests pass on your pt/xla PR, land that PR (pt/xla master starts failing)
(3) update your core PR’s `xla.txt` file (again) to point to the commit hash from tip-of-master pt/xla (in general when you do this, should you wait to confirm green CI? Or just insta-force-land. I kind of think we should just force land assuming CI was green the first time around.
```

I think pytorch pr can skip (1) if I verify on xla end all test is passing, but ideally this pr should be rebased before XLA pr went in. The goal is that pytorch pr update `xla.txt` right after xla pr merge to reduce the xla side CI churn.

**Comment by JackCaoG** [comment]
Test is green on the xla pr https://github.com/pytorch/xla/pull/3554. We can coordinate a land tmr.

**Comment by JackCaoG** [comment]
XLA pr is ready.

**Comment by JackCaoG** [comment]
so let's do following things.
1. rebase this pr(otherwise there will be merge conflict when you edit the `xla.txt`.
2. let me know when you are ready
3. I will merge pytorch/xla pr. Then you have a xla hash (the pytorch/xla head commit hash)
4. you update `xla.txt` and force merge without waiting for the test

**Comment by kurtamohler** [comment]
@JackCaoG , PR is rebased. I can update `xla.txt` when the hash is ready

**Comment by JackCaoG** [comment]
@kurtamohler sg, just merged my pt/xla pr. Please update the hash to `35f759fdd7eb585679df7c1e6db4569b1aba5475`

**Comment by kurtamohler** [comment]
Thanks @JackCaoG !

**Comment by kurtamohler** [comment]
I'm not sure what's up with the `pull / linux-bionic-py3.7-clang9 / build` CI job. It just says "This check failed" with no other information and I don't see any log output

**Comment by bdhirsh** [comment]
force merging to unblock XLA CI. The failing CI looks like it was passing before the last update.

**Comment by malfet** [comment]
@kurtamohler you need to update kernel signature for `torch.sum` for MPS as well

**Comment by JackCaoG** [comment]
I will revert XLA pr too

**Comment by zou3519** [comment]
@kurtamohler: before you attempt another merge: this PR will definitely break the functorch fb-internal build. This is currently grounds for revert (but we're working on making the experience better). We should coordinate a change to ensure that functorch doesn't break
